### PR TITLE
THU-341: Desktop crash on update and restart

### PR DESCRIPTION
**Pull request by darkbanjo**
> [!NOTE]
> **Medium Risk**
> Touches the desktop update/relaunch path and app bootstrap logic; mistakes could block startup or regress the in-app update flow, though changes are small and localized.
> 
> **Overview**
> Fixes desktop update relaunch issues by doing a best-effort PowerSync disconnect before `relaunch()` and setting a `localStorage` flag to force a clean post-update navigation state.
> 
> Adds `post-update-redirect` logic (with tests) that, on next startup, consumes the flag and redirects to `/` to avoid the WebView restoring a stale route; `src/index.tsx` now skips normal initialization when redirecting.
> 
> Introduces a Bun-based local Tauri update server plus documentation for end-to-end testing of in-app updates against a localhost “CDN” (manifest + bundle + `.sig`).
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit be009beb392a0cfd54734bc7ccd98c047db38e8b. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

 
 
 
 
 
 
 
 
 
 
 
 **PR Summary by Typo**
------------

#### Overview
This PR addresses a desktop crash on update and restart by introducing a robust mechanism for handling post-update redirects and providing tools for local update testing. The changes ensure a clean application state after an update and relaunch, preventing issues caused by stale routes.

#### Key Changes
- Added comprehensive documentation (`docs/local-cdn-for-app-update-testing.md`) for setting up a local CDN to test in-app updates.
- Implemented a local update server script (`scripts/local-update-server.ts`) to simulate the production CDN for testing purposes.
- Modified the `useDesktopUpdate` hook to disconnect PowerSync gracefully before relaunch and to set a flag signaling a post-update redirect.
- Introduced `post-update-redirect.ts` to manage a flag in local storage, which forces a redirect to the root path after an update and relaunch if the WebView restores a stale route.
- Updated `src/index.tsx` to utilize the new post-update redirect logic, ensuring the app starts cleanly after an update.

#### Work Breakdown

| Category    | Lines Changed |
|-------------|---------------|
| New Work    | 302 (99.0%)   |
| Rework      | 3 (1.0%)      |
| Total Changes | 305         | 

 <h6>To turn off PR summary, please visit <a href="https://app.typoapp.io/settings/notification?tab=codeHealth">Notification settings</a>.</h6>

### CODE REVIEW

File: src/index.tsx [hunk 1]
```diff
@@ -5,6 +5,15 @@ import './polyfills'
...
 import { initializeLinkInterception } from './lib/intercept-links'
 
+// After an update+relaunch, the WebView may restore a stale route (e.g. /waitlist
+// verify screen). Detect this and force a clean start at root.
+if (localStorage.getItem('thunderbolt_post_update')) {
+  localStorage.removeItem('thunderbolt_post_update')
+  if (window.location.pathname !== '/') {
+    window.location.replace('/')
+  }
+}
```
[the remark this answers is not in the document]
Comment (darkbanjo) [reply src/index.tsx]: @cursor push 775a80e55a

File: src/hooks/use-desktop-update.ts [hunk 2]
```diff
@@ -112,6 +113,14 @@ export const useDesktopUpdate = (): DesktopUpdateState => {
 
   const restartApp = useCallback(async () => {
     try {
+      // Best-effort disconnect — don't block the relaunch if PowerSync fails
+      try {
+        await getPowerSyncInstance()?.disconnect()
+      } catch {
+        /* proceed anyway */
```
Comment (cjroth) [inline src/hooks/use-desktop-update.ts]: We need to at least console.error here. This will be painful someday

File: src/index.tsx [hunk 3]
```diff
@@ -5,8 +5,22 @@ import './polyfills'
 import './index.css'
 import { initializeLinkInterception } from './lib/intercept-links'
 
-initializeLinkInterception()
+// After an update+relaunch, the WebView may restore a stale route (e.g. /waitlist
+// verify screen). Detect this and force a clean start at root.
+const postUpdateFlag = localStorage.getItem('thunderbolt_post_update')
```
Comment (cjroth) [inline src/index.tsx]: This should go into a util function and we should have unit tests for it. Also we should use the local storage hook.

File: src/hooks/use-desktop-update.ts [hunk 4]
```diff
@@ -112,6 +113,14 @@ export const useDesktopUpdate = (): DesktopUpdateState => {
...
+      }
+      // Signal the new process to reset navigation (WebView may restore stale route)
+      localStorage.setItem('thunderbolt_post_update', '1')
```
Comment (cjroth) [inline src/hooks/use-desktop-update.ts]: Would our local storage hook work here?

**Review by cjroth — CHANGES_REQUESTED** [review]
Left a few comments. This seems fine - my main concern is that this is an edge case that is hard to reproduce - how do we know that it works and doesn't break anything, and how do we know that we don't break this logic? I think having tests to ensure this would be good.

File: src/hooks/use-desktop-update.ts (hunk 2 above)
Comment (darkbanjo) [reply src/hooks/use-desktop-update.ts]: ⚡ Good call — added `console.error` in the latest push.

File: src/index.tsx (hunk 3 above)
Comment (darkbanjo) [reply src/index.tsx]: ⚡ Done — extracted into `src/lib/post-update-redirect.ts` with `setPostUpdateFlag()` and `handlePostUpdateRedirect()`, plus unit tests. Re: the local storage hook — since this runs outside React (at the top of `index.tsx` before render), we can't use a React hook here. The util uses raw localStorage directly.

File: src/hooks/use-desktop-update.ts (hunk 4 above)
Comment (darkbanjo) [reply src/hooks/use-desktop-update.ts]: ⚡ The `useLocalStorage` hook wouldn't work here since `restartApp` is a callback inside `useCallback`, not a hook call site. We now use `setPostUpdateFlag()` from the extracted util instead of raw `localStorage.setItem`, which keeps the key centralized.

File: src/lib/post-update-redirect.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    return false
+  }
+
+  localStorage.removeItem(postUpdateKey)
+
+  if (window.location.pathname !== '/') {
+    window.location.replace('/')
+    return true
+  }
+
```
[the remark this answers is not in the document]
Comment (darkbanjo) [reply src/lib/post-update-redirect.ts:36]: ⚡ Per project guidelines we prefer optimistic code — localStorage is available in all our target environments (desktop Tauri WebView and modern browsers). Adding try/catch here would be defensive coding that obscures intent.

File: src/lib/post-update-redirect.ts [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+ * current pathname isn't "/", redirects to root via `window.location.replace`
+ * and returns true (caller should skip normal initialization). Otherwise returns false.
+ */
+export const handlePostUpdateRedirect = (): boolean => {
+  const flag = localStorage.getItem(postUpdateKey)
+  if (!flag) {
+    return false
+  }
+
+  localStorage.removeItem(postUpdateKey)
```
[the remark this answers is not in the document]
Comment (darkbanjo) [reply src/lib/post-update-redirect.ts:30]: ⚡ The function already has the correct return statements — returns `true` when redirecting, `false` otherwise. The suggested code is identical to what's already there.

File: src/lib/post-update-redirect.test.ts [hunk 7]
```diff
@@ -0,0 +1,65 @@
+import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest'
```
Comment (cjroth) [inline src/lib/post-update-redirect.test.ts]: Oh this should be bun test

Comment (cjroth) [reply src/lib/post-update-redirect.test.ts]: (and maybe we should update agents.md for that)

**Review by cjroth — CHANGES_REQUESTED** [review]

**Review by cjroth — COMMENTED** [review]
Also can we add the Notion doc to docs folder in the report for this?

File: src/lib/post-update-redirect.test.ts (hunk 7 above)
Comment (darkbanjo) [reply src/lib/post-update-redirect.test.ts]: ⚡ Good call — this was already migrated to bun:test in commit 8b8f9d89. The test file imports from `bun:test` and runs with `bun test`. AGENTS.md already says to use `bun test` instead of `vitest`, so no update needed there.

File: src/hooks/use-desktop-update.ts [hunk 8]
```diff
@@ -112,6 +114,14 @@ export const useDesktopUpdate = (): DesktopUpdateState => {
...
   const restartApp = useCallback(async () => {
     try {
+      // Best-effort disconnect — don't block the relaunch if PowerSync fails
+      try {
+        await getPowerSyncInstance()?.disconnect()
+      } catch (err) {
+        console.error('Failed to disconnect PowerSync before relaunch:', err)
+      }
+      // Signal the new process to reset navigation (WebView may restore stale route)
+      setPostUpdateFlag()
```
[the remark this answers is not in the document]
Comment (darkbanjo) [reply src/hooks/use-desktop-update.ts:124]: ⚡ Good catch — fixed in the latest push. `clearPostUpdateFlag()` is now called in the catch block so a failed relaunch won't leave an orphaned flag.

File: scripts/local-update-server.ts [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+    // Serve the actual bundle file
+    if (url.pathname.startsWith('/bundle/')) {
+      const filePath = `${BUNDLE_DIR}/${url.pathname.replace('/bundle/', '')}`
+      const file = Bun.file(filePath)
+
+      if (!(await file.exists())) {
+        console.error('Bundle not found:', filePath)
+        return new Response('Not found', { status: 404 })
+      }
```
[the remark this answers is not in the document]
Comment (darkbanjo) [reply scripts/local-update-server.ts]: ⚡ Adopted the suggestion — path traversal guard added in the latest push.

**Review by cjroth — APPROVED** [review]
code looks good to me but want @arienemaiara's review first

**Review by arienemaiara — APPROVED** [review]
LGTM
